src/main/java/com/example/demo/utils/DefaultPushMessageEncryptionUtil.java

```diff
@@ -0,0 +1,203 @@
+			hmacSHA256 = Mac.getInstance("HmacSHA256");
+		    hmacSHA256.init(new SecretKeySpec(Base64.getUrlDecoder().decode(authSecret), "HmacSHA256"));
+		} catch (NoSuchAlgorithmException | InvalidKeyException e1) {
+			throw new InvalidDataException();
+		}
+	    byte[] prkKey = hmacSHA256.doFinal(ecdhSecret);
+
+	    byte[] keyInfo = concat("WebPush: info\0".getBytes(StandardCharsets.UTF_8),uncompressECPublicKey(uaPublicKey), uncompressedASPublicKey);
+	    
+	    try {
+			hmacSHA256.init(new SecretKeySpec(prkKey, "HmacSHA256"));
+		} catch (InvalidKeyException e1) {
+			throw new InvalidDataException();
+		}
+	    hmacSHA256.update(keyInfo);
+	    hmacSHA256.update((byte) 1);
+	    byte[] ikm = hmacSHA256.doFinal();
+
+	    try {
+			hmacSHA256.init(new SecretKeySpec(salt, "HmacSHA256"));
+		} catch (InvalidKeyException e1) {
+			throw new InvalidDataException();
+		}
+	    byte[] prk = hmacSHA256.doFinal(ikm);
+
+	    byte[] cekInfo = "Content-Encoding: aes128gcm\0".getBytes(StandardCharsets.UTF_8);
+
+	    try {
+			hmacSHA256.init(new SecretKeySpec(prk, "HmacSHA256"));
+		} catch (InvalidKeyException e1) {
+			throw new InvalidDataException();
+		}
+	    hmacSHA256.update(cekInfo);
+	    hmacSHA256.update((byte) 1);
+	    byte[] cek = hmacSHA256.doFinal();
+	    cek = Arrays.copyOfRange(cek, 0, 16);
+
+	    byte[] nonceInfo = "Content-Encoding: nonce\0".getBytes(StandardCharsets.UTF_8);
+
+	    try {
+			hmacSHA256.init(new SecretKeySpec(prk, "HmacSHA256"));
+		} catch (InvalidKeyException e1) {
+			throw new InvalidDataException();
+		}
+	    hmacSHA256.update(nonceInfo);
+	    hmacSHA256.update((byte) 1);
+	    byte[] nonce = hmacSHA256.doFinal();
+	    nonce = Arrays.copyOfRange(nonce, 0, 12);
+
+	    Cipher cipher;
+		try {
+			cipher = Cipher.getInstance("AES/GCM/NoPadding");
+		    cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(cek, "AES"),new GCMParameterSpec(128, nonce));
+		} catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | InvalidAlgorithmParameterException e1) {
+			throw new InvalidDataException();
+		}
+
+	    List<byte[]> inputs = new ArrayList<>();
+	    byte[] plainTextBytes = plainText.getBytes(StandardCharsets.UTF_8);
+	    inputs.add(plainTextBytes);
+	    inputs.add(new byte[] { 2 }); // padding delimiter
+
+	    int padSize = Math.max(0, PADDING_SIZE_0 - plainTextBytes.length);
+	    if (padSize > 0) {
+	      inputs.add(new byte[padSize]);
+	    }
+
+	    byte[] encrypted;
+		try {
+			encrypted = cipher.doFinal(concat(inputs.toArray(new byte[0][])));
+		} catch (IllegalBlockSizeException | BadPaddingException e) {
+			throw new InvalidDataException();
+		}
+
+	    ByteBuffer encryptedArrayLength = ByteBuffer.allocate(4);
+	    encryptedArrayLength.putInt(encrypted.length);
+
+	    byte[] header = concat(salt, encryptedArrayLength.array(),
+	        new byte[] { (byte) uncompressedASPublicKey.length }, uncompressedASPublicKey);
+	    
+	    return concat(header, encrypted);
+	}
```

Comment:
can we split this method into smaller ones that will have descriptive names so that logic is easier to follow